### DISCUSSION TITLE
Saving/Loading presets of my VST

### DISCUSSION THREAD

**Discussion by dervur** [RomanKubiak/ctrlr#662] [Q&A] [answered]
Hello there,

I searched left and right these discussions and the original forum (I gave up to that really quick as I can't create a new account to download files apparently). But I can't seem to find an answer to my question. 

I'm trying to think how to save and load presets of my panels in a way that might work for standalone or VST instances. A way to save a file externally that has the information of all the modulator values. This file can then be loaded in my restricted instance to set the values of all my modulators. I know how to set a value by LUA. Still did not find a way to get a value using LUA (having issues in https://github.com/RomanKubiak/ctrlr/issues/661), pleas help me with this too. 
But then how would I save this outside and load it from there? Is there a simple way to do this? 

The way I was picturing it is by saving a string of vstIndex=value in a txt file divided by spaces that can be read and saved easily with a for loop.
Then I saw that by setting the value of a modulator a MIDI out is sent as well, so this should solve the next issue which is communicate to the device this preset.

Does anybody know a way to do what I need to do? If there is already a discussion on this I might've missed please let me know, I might've not found it due to a difference in terminology for example.

**Comment by damiensellier** [comment]
I invite you to take a look at available panels on ctrlr.org who are labelled as librarian like the ms2000 or Novation x station orJD990 :
https://ctrlr.org/rems2000-korg-ms2000-editor/
[Novation X-Station Voice Editor/](https://ctrlr.org/novation-x-station-voice-editor/)
https://ctrlr.org/roland-jd990-super-editor/

They have everything under the hood to learn how to deal with valueTree, XML and the JUCE File class.

hope it helps
Damien

**Reply by dervur** [reply]
Hi Damien thank you! Some from Ctrlr.org do not work in the new version anymore (like the korg ms2000 you sent me). Novation X is huge so I'm trying to delve into all the feature. Thank you very much. Currently I'm looking at something more simple like the DEMO FIles panel you created long ago.
The save by text feature even on your panel does not work (it makes Ctrlr crash) I don't know if it just a new version of Ctrlr issue. The save by byte works well and I understand it easily so I think I will go with that for my model. 

This brings to the next question, the idea I had was to save bytes as VstIndex followed by value, so I can gather and save all my modulators values and and gather the data using for loops. The issue is that I found out that I cannot get Modulator by VstIndex via lua. To do that should I modify CtrlrLuaManager.cpp at around line 860ish with something like `def("getModulatorByVstIndex", &CtrlrPanel::getModulatorByVstIndex)`? And do a build with that?
Because I saw there is a "getModulatorByVstIndex" in CtrlrManager.cpp already. As my programming skills are not excellent correct me if I'm wrong please. I wanted to do a for loop to get vstIndex followed by value.

**Reply by damiensellier** [reply]
DEMO panels are not from me, they are from Roman Kubiak the creator of ctrlr. 
The newer version 5.6.30 is from me but there is not much difference concerning LUA scripts. 
Usually the old panels crash because of lua scripts including send midi message with the older syntax.
you can study any panels with ctrlr 5.2.198 they won't crash. 

The method to save values would be to run a loop through all modulators, create a table, get their value, place the values in the table, convert the table to memoryblock and save the MB as file.
This is exactly what dnaldoog proposed recently for the Akai AX80 panel : https://github.com/RomanKubiak/ctrlr/discussions/645

https://github.com/RomanKubiak/ctrlr/files/15493348/AX80_2_0.zip

you have everything needed there ;)

Good luck!

EDIT: Link to panel. updated

**Reply by dervur** [reply]
Hi thank you, the issue is that I'm not much of a programmer and I'm a bit confused by the whole sysex thing. I will have to read more about it but I wanted for now a simple solution. Anyway I found a way to save sysex data based on that panel and then convert it. I will post what I did below as answer for other people that come across this thread and want a simple quick solution.

Thank you very much again for your help, you sent me a lot of material that I now need to study hahaha

**Comment by dnaldoog** [comment]
JSON is the way to go!
[example panel here](https://sourceforge.net/projects/ctrlr-utilities/files/developer%20tools/)

**Reply by dervur** [reply]
The issue is that I'm not really good at programming so I got a little scared of the JSON thing when I saw it on another thread. I will surely consider it for the future, for now I'm saving easy hex data of my panel, I will post what I did below for other people that come across this thread.
Thank you very much anyway for positing this I will surely look into it!

**Comment by dervur** [comment, marked as answer]
Ok, this is how I save them for other people coming here. I will post the one for the panel with only 10 values, but I used the same approach for another one that has 162 modulators.

First of all you need a function to get the panel data:
```
function getPanelData()
    local data = {}

    -- Populate the data table
    data[1] = panel:getModulatorByName("Pulse_Colour"):getValue()
    data[2] = panel:getModulatorByName("Pulse_Level"):getValue()
    data[3] = panel:getModulatorByName("Amp_Attack"):getValue()
    data[4] = panel:getModulatorByName("Amp_Decay"):getValue()
    data[5] = panel:getModulatorByName("Drive"):getValue()
    data[6] = panel:getModulatorByName("Tone"):getValue()
    data[7] = panel:getModulatorByName("Resonator_Pitch"):getValue()
    data[8] = panel:getModulatorByName("Resonator_Bend"):getValue()
    data[9] = panel:getModulatorByName("Resonator_Time"):getValue()
    data[10] = panel:getModulatorByName("Accent"):getValue()
    
    return data
end
```
If from the modulator list you export a csv, you can ask copilot or chat gpt to help you code this, just write the first line and ask AI to expand the code for the other modulators (you can paste them)
Then the script to save the preset, I decided to create a custom extension .kvlk that did not exist before:
```
function savePreset()

	-- Default filename
	local cPresName = "preset"

	-- File to save
	local file = utils.saveFileWindow("Save Volca Kick preset to file...", File(cPresName), "*.kvlk", true)

	-- Writing data to the file
	file:replaceWithData(MemoryBlock(getPanelData()))

end
```
For getting the data the thinfs are a bit more complicated. When you read the file you will have an hex string, so first of all let's do a function that converts this hex string to a table for easy recall:
```
function hex_to_table(hex_string)
    local data = {}
    for hex in string.gmatch(hex_string, '%S+') do
        table.insert(data, hex)
    end
    return data
end
```
And this is the code needed to set the panel data
```
function setPanelData(data)

    -- Create a table with modulator names and numbers set in relation to the index of the getPanelData function
	local modulator = {}
    modulator["Pulse_Colour"] = 1
    modulator["Pulse_Level"] = 2
    modulator["Amp_Attack"] = 3
    modulator["Amp_Decay"] = 4
    modulator["Drive"] = 5
    modulator["Tone"] = 6
    modulator["Resonator_Pitch"] = 7
    modulator["Resonator_Bend"] = 8
    modulator["Resonator_Time"] = 9
    modulator["Accent"] = 10


    -- Set values
	for k,v in pairs(modulator)do
        panel:getModulatorByName(k):setValue(data[v], true)
	end
end
```
Lastly, the load preset function, I'll probably expand it to handle more error, but this works fine, I force the open dialog to only open kvlk files:
```
function loadPreset()
	-- File open dialog
    local file = utils.openFileWindow(
		"Select Volca Kick preset...",
		File.getSpecialLocation(File.userHomeDirectory),
		"*.kvlk",
		true
	)

	local hexData = MemoryBlock()

	-- If file exists, then proceed
	if file:existsAsFile() then
		hexData = MemoryBlock(file:getSize())
		file:loadFileAsData(hexData)
	else
		return
	end

	-- If the file is too big than there is a mistake
	if (file:getSize() > 10) then
		utils.warnWindow("Warning", "Wrong file size, operation cancelled")
	else
		-- Convert hex data into table
		local data = hex_to_table(hexData:toHexString(1))
        -- Convert hex to decimal number
        local dec_data = {}
        for i, v in ipairs(data) do
            dec_data[i] = tonumber(v,16)
        end
		-- Call function to set values in the VST
		setPanelData(dec_data)
	end

	-- Get filename and change extension
    filename = string.gsub(file:getFileName(), "(.+)%.%w+$", "%1")
    -- Set filename in the LCD screen
    panel:getComponent("Preset_Name"):setPropertyString("uiLabelText", filename)

end
```

Hopefully this helps people approaching Ctrlr and have not a strong programming background. Copilot helped me a lot in doing what I didn't know how to, especially since I never touched lua before :)

**Comment by dnaldoog** [comment]
Here is how I would do that.


The JSON code is much simpler, although I acknowlege it's yet another thing to knowledge, but if you have  a straight one dimensional **hash/value** table as below, it's actually much simpler and you don’t really need to know what JSON is!

<pre>
--
-- Called when a mouse is down on this component
--

loadPreset = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    -- File open dialog
    local jsonfile =
        utils.openFileWindow(
        "Select Volca Kick preset...",
        File.getSpecialLocation(File.userHomeDirectory),
        "*.kvlk",
        true
    )

    if jsonfile:existsAsFile() then
        local str = jsonfile:loadFileAsString()
        obj = json.decode(str)
    end
    if type(obj) ~= "table" then
        return
    end -- bail out if something's not right
    for k, v in pairs(obj) do
        if data[k] then -- safeguard against non existent keys
            panel:getModulatorByName(k):setModulatorValue(v, false, true, false)
        end
    end
end



--
-- Called when a mouse is down on this component
--

savePreset = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    tmp = {} -- this table will converted to JSON and saved to file
    for key in pairs(data) do -- assign values to each item
        tmp[key] = panel:getModulatorByName(key):getModulatorValue()
    end
    -- Default filename
    local cPresName = "preset"

    -- File to save
    local jsonfile = utils.saveFileWindow("Save Volca Kick preset to file...", File(cPresName), "*.kvlk", true)
    -- Writing data to the file
    jsonfile:replaceWithText(json.encode(tmp), true, true)
end


data = {
    Pulse_Colour=true,
    Pulse_Level=true,
    Amp_Attack=true,
    Amp_Decay=true,
    Drive=true,
    Tone=true,
    Resonator_Pitch=true,
    Resonator_Bend=true,
    Resonator_Time=true,
    Accent=true,
} 
-- just add new modulators to this list (truth set)
-- if your modulator name contains a space or funny character
-- surround by [''] e.g. ['Pulse Level']=true,
</pre>

[Save import-export JSON_1_0.zip](https://github.com/user-attachments/files/16128316/Save.import-export.JSON_1_0.zip)

**Reply by dnaldoog** [reply]
P.S. In the original message I wrote why indexing your modulators numerically is not such a good idea, but there _are_ arguments for it as well, so please just ignore my ramblings; still I think the JSON key/value way is easier to maintain down the line. 😊

**Reply by dervur** [reply]
Thank you for that I will take a look surely!
I am aware that the why indexed it is horrible from a programming perspective (you better do not see the code for the other panel or you might have stroke hahaha). But I do not think I will include anything else on the panel or modify what is already there, that's why I chose the easy programming route instead of the smart one... Ok the fact the Copilot wrote most of the repetitive code for me is the reason as well especially for the other 162 values panel

The reality is that last week I got down a rabbit hole in trying to solve this issue, now I want to leave it as it is cause I spent too much time on this and ignored too many other things :)
But for the future I will surely take a look on json and I am truly grateful for you posting this!

**Comment by dnaldoog** [comment]
Oh no, it's not horrible from a programming perspective and if it works then it's fine. I just think when you come back to it (which you may or may not) it might be harder to unravel what was going on, or maybe not. - so if it works then great!

Spending too much time on Ctrlr? It's the story of my life ! :)

**Reply by dervur** [reply]
Trust me, the other panel is horrible... It is basically the copy of 16 modulators for 10, and they all end per "Part-1", "Part-2" ... "Part-10"... but the for loop I initially wrote to iterate through them and their names was not working so now I got more than 160 lines with if/elseif statements hahahaha (obvsiuly with the help of copilot)... That I consider horrible for real hahaha

Anyway as I'm adding more code to handle errors I'm adding more comments too, at least I have the habit to use comments

I had to clean my flat last week and I didn't because I was too absorbed, that's why I want a break for some time hahaha

**Comment by dnaldoog** [comment]
I looked at your binary file version. There are a couple of improvements that can be made. 

You don't need that fancy *hex_to_table()* function, because MemoryBlock contains a method to do just that **toLuaTable()** - and also I would keep all access to getting/setting the modulators in one indexed table **modulator={}**: 

<pre>
-- create a global table for all reference to modulators
-- DRY principle
modulator = {
   "Pulse_Colour",
   "Pulse_Level",
   "Amp_Attack",
   "Amp_Decay",
   "Drive",
   "Tone",
   "Resonator_Pitch",
   "Resonator_Bend",
   "Resonator_Time",
   "Accent",
}

function getPanelData()
    local t = {}
	for i,v in ipairs(modulator) do
	table.insert(t,panel:getComponent(v):getValue())
	end
    -- Populate the data table
return t
end

function savePreset()
	-- Default filename
	local cPresName = "preset"
	-- File to save
	local file = utils.saveFileWindow("Save Volca Kick preset to file...", File(cPresName), "*.kvlk", true)
	-- Writing data to the file
	file:replaceWithData(MemoryBlock(getPanelData()))
end

function loadPreset()
	-- File open dialog
    local file = utils.openFileWindow(
		"Select Volca Kick preset...",
		File.getSpecialLocation(File.userHomeDirectory),
		"*.kvlk",
		true
	)

-- MemoryBlock has a method for converting binary strings to lua table
-- hexData:toLuaTable(t)
	local hexData = MemoryBlock()
	local t={}
	-- If file exists, then proceed
	if file:existsAsFile() then
		hexData = MemoryBlock(file:getSize())
		file:loadFileAsData(hexData)
	else
		return
	end

	-- If the file is too big than there is a mistake
	if (file:getSize() > 10) then
		utils.warnWindow("Warning", "Wrong file size, operation cancelled")
	else
		-- Convert hex data into table
		hexData:toLuaTable(t)
        -- Convert hex to decimal number
          for i, v in ipairs(t) do
		-- Call function to set values in the VST
		panel:getModulatorByName(modulator[i]):setModulatorValue(v,false,true,false)
		end
	end
	-- Get filename and change extension
    filename = string.gsub(file:getFileName(), "(.+)%.%w+$", "%1")
    -- Set filename in the LCD screen
    panel:getLabelComponent("Preset_Name"):setText(filename)
</pre>

[Save import-export-BINARY_LOAD.zip](https://github.com/user-attachments/files/16152523/Save.import-export-BINARY_LOAD.zip)

**Reply by dervur** [reply]
Oh thank you very much I did not know that a function for that exists. Thank you!
